syntax = "proto3";

package ProtoSpace;

message SlashFilter{
    bool  use_slash_filter                           = 1;
    int32 slash_filter_level                         = 2;
    bool  exclude_slash                              = 3;
    bool  include_slash                              = 4;
    bool  enable_exceptions                          = 5;
    repeated int32 fandom_exceptions                 = 6; 
}

message Recommendations{
    bool list_open_mode                             = 1;
    int32 list_for_recommendations                  = 2;
    int32 use_this_recommender_only                 = 3;
    int32 min_recommendations                       = 4;
    bool show_origins_in_lists                      = 5;
}

message BasicFilters{


    string website                  = 1;
    int32 min_words                 = 2;
    int32 max_words                 = 3;
    int32 max_fics                  = 4;
    int32 min_favourites            = 5;
    bool allow_unfinished           = 6;
    bool allow_no_genre             = 7;
    bool ensure_completed           = 8;
    bool ensure_active              = 9;
}

message ReviewBias{
    enum ReviewBiasMode{
        bias_none           = 0;
        bias_favor          = 1;
        bias_exclude        = 2;
    };
    enum BiasOperator{
        bo_more             = 0;
        bo_less             = 1;
    };

    ReviewBiasMode bias_mode   = 1;
    BiasOperator bias_operator = 2;
    double bias_ratio          = 3;
}

message RecentAndPopular{
    int32  fav_ratio   = 1;
    string date_cutoff = 2;
}


message ContentMatchFilter
{
    string fandom                       = 1; 
    bool include_crossovers             = 2;
    bool crossovers_only                = 3;
    bool other_fandoms_mode             = 4;
    
    repeated string genre_exclusion     = 5;
    repeated string genre_inclusion     = 6;
    repeated string word_exclusion      = 7;
    repeated string word_inclusion      = 8;
}

message TagFilter{
    bool ignore_already_tagged  = 1;
    repeated string active_tags = 2;
    repeated int32 all_tagged   = 3; // will see how this affects perfomance 
}

message SizeLimits{
    int32 record_limit              = 1;
    int32 record_page               = 2;
}

message Filter{
    enum FilterMode{
        filtering_undefined          = 0;
        filtering_in_fics            = 1;
        filtering_in_recommendations = 2;
        filtering_whole_list         = 3;
    };
    enum SortMode {
    sm_undefined         = 0;
    sm_favourites        = 1;
    sm_fav_rate          = 2;
    sm_update_date 	     = 3;
    sm_publish_date      = 4;
    sm_rec_count         = 5;
    sm_wcr_cr            = 6;
    sm_rev_to_fav        = 7;
    sm_genre_values      = 8;
    sm_wordcount         = 9;
  }
    
    bool randomize_results                      = 1;
    BasicFilters basic_filters                  = 2;
    FilterMode filtering_mode                   = 3;
    SizeLimits size_limits                      = 4;
    SortMode sort_mode                          = 5;
                
    ReviewBias review_bias                      = 6;
    RecentAndPopular recent_and_popular         = 7;
    ContentMatchFilter content_filter           = 8;
    TagFilter tag_filter                        = 9;
    SlashFilter slash_filter                    = 10;
    Recommendations recommendations             = 11;
}
